Improved cli and misc maintenance

ptsOSL opened the pull request:
- Updated the CLI to allow configuration of a time step delay and a filepath to save data at. 
- Updated the way CA ports are configured, these are now set from your environment variables
- Fixing the Dockerfile to allow this module to be run as a container for the system tests
- Various other fixes, such as logging changes and getting the ringmode

ptsOSL committed:
Update gitignore

When run with the default filepath, this module creates
response matrix build directories in the top level
of this directory. These do not want to be commited
to the repo, so we now ignore them.

ptsOSL committed:
Replace deprecated 'true' option with 'explicit'

The explicit option will reorganise imports only when a file
is manually saved. https://code.visualstudio.com/docs/editing/refactoring

ptsOSL committed:
Install libqt5gui5 to fix missing dependency in devcontainer

Without installing this, running dls-response-matrix fails with error
libGL.so.1: cannot open shared object file: No such file or directory

ptsOSL committed:
Set default CA ports to simulation ports

By default the app should run against the simulated machine, this
behaviour can be overwritten by passing environment variables when
running the container against the actual machine.

ptsOSL committed:
Remove DISPLAY env variable from devcontainer

This is no longer needed and was causing the app to fail to connect
to the x11 display. This variable is now automatically set by the devcontainer
extension to the correct value.

ptsOSL committed:
Reorganise devcontainer settings option

Moved this to the new recommended location as the old location
is deprecated.

ptsOSL committed:
Improve how ca ports are configured

Previously we attempted to set the CA addr and repeater ports
from within python. The preferred way is to read the current
values from the users environment. We now do this and raise
an error or log a warning if the variables are incompatible
with the args supplied to the app.

ptsOSL committed:
Update CA port configuration tests

Update to tests so they pass against the new way of
configuring the CA repeater and server ports. Make the tests
check that an exception is raised incase of invalid ports.

ptsOSL committed:
Make the tests run up against simulation CA ports

It is assumed that most of the time these tests will be
run against the virtual accelerator on ports 806X. This configuration
prevents accidentally running the tests against he real machine.

ptsOSL committed:
Add filepath as CLI argument

ptsOSL committed:
Add logic for handling new filepath cli option

Instead of saving data to cwd, we check if the cli argument
filepath is None, if it is not, then we try and use the value
passed and raise an exception if the filepath is invalid. If it
is None then we write to cwd.

ptsOSL committed:
Change results to use filepath from the updated config

Previously the from_csv function was passed a filepath, now the filepath
configured in the Config class is used here and elsewhere. When writing the
csv we also use the filepath from the Config class if defined, else cwd.

ptsOSL committed:
Update tests to work with latest filepath changes

ptsOSL committed:
Add filebrowser to gui to allow filepath selection

ptsOSL committed:
Add cli option for step delay

ptsOSL committed:
Add logic for handling custom step delays

This includes checking values are within boundaries

ptsOSL committed:
Update tests for new step delay option

ptsOSL committed:
Add step delay and max/min bounds

You can now specify the step delay from the gui. Also
added are checks that the step delay and step delta are
within the range of valid values which depends on the machine
type selected.

ptsOSL committed:
Improve handling of ringmode lookup

Give an extra error message suggesting that slow-orbit-feedbacks may
not be running before re-raising.

(cherry picked from commit c6908b3cec43cbd8fb7dc1bd13edb2d95c48fd18)

ptsOSL committed:
Reset logger before applying logging config

This prevented getting the file logging strings duplicated to the
console.

MJGaughran commented on src/dls_response_matrix/response_matrix.py:
```diff
@@ -31,13 +34,24 @@ def get_ring_modes() -> tuple[list, str]:
+        ).enums
+        current_ringmode = caget("SR-CS-RING-01:MODE", datatype=str, throw=True)
+    except ca_nothing as e:
+        ring_mode_list = ["I04"]
```
Is there any way to provide a default ring mode that isn't hard-coding it in the middle of get_ring_modes()?

MJGaughran replied:
At the very least, it should be a 'constant' e.g. DEFAULT_RING_MODE at the top of the module

MJGaughran replied:
But why not use the ring mode specified in the configuration?

ptsOSL replied:
We now use DEFAULT_MACHINE_MODE. Updated as part of https://github.com/DiamondLightSource/dls-response-matrix/pull/33/commits/c6a86a0dce451f695ec3ebafcf2e49564908f10d

MJGaughran commented on tests/test_results.py:
```diff
@@ -65,39 +69,51 @@ def test_results_remove_bpms_works_as_expected():
 
 
 def test_from_csv_raises_exception_if_new_filename_same_as_old_filename(tmp_path):
+    """Create a metadata json file using the config supplied. Create a dummy matrix and
```
For both of these tests, I'm not sure the extended description is too helpful as you can see the code immediately below. If a particular step isn't clear, feel free to include a comment on that particular line.

ptsOSL replied:
I have removed this as I dont think its needed anymore, some clarifying comments have been added to test body. See https://github.com/DiamondLightSource/dls-response-matrix/pull/33/commits/1a725e6989a257e8f26fe42f2431918febb3abcc

ptsOSL committed:
Improve handling of slow feedbacks IOC being down

We now give a warning when the ring mode is unavailable instead of an exception.
> An exception is only re-raised when the user attempts to use the disable
> correctors feature which requires the slow feedbacks IOC. This makes the
> behavior of the CLI and GUI the same and the slow feedbacks IOC is now only
> required when trying to disable correctors.

ptsOSL committed:
Improve handling of filepath/name in Results class

Each instance of the results class now stores its own filename and
filepath which is the only filename/filepath used to save data. These
are either set from a passed Config object or set directly when loading
a csv file. This allows us to create an initial results object and
directory and also later create a new Results object and directory in a
new location with a new name using the same code.

ptsOSL committed:
Update README

Add advice for using the module

MJGaughran reviewed (approved):
We've agreed to set up Copier template and working CI/CD in a separate PR, as it will leave the repository in a decent state for future work.
